packages/javascript-decorators/src/memoize.ts

```diff
@@ -0,0 +1,73 @@
+export interface MemoizeMap<T, U> {
+  get(key: T): U;
+  has(key: T): boolean;
+  set(key: T, value: U): MemoizeMap<T, U>;
+}
+
+type Resolver<T extends Function> = T extends (...args: infer T) => any
```

Comment:
This PR is inspiring me to read up on advanced types 😬 😆

Reply:
I found this in a github comment!
This is how you enforce the parameters type using the type of a function.
The post also mentioned something similar to enforce the return type as well.

Thou I read in Chris's PR yesterday, `ReturnType` is actually a thing people use so much that it is included in the base library!

Reply:
Don't use `infer T` though, since `T` is already the generic name of the function type (generally, this ends up being called `U`, but you can call it `Arguments` or something more meaningful

Reply:
I may need an example, not sure I know exactly what you mean.

Reply:
```ts
type Resolver<T extends Function> = T extends (...args: infer Arguments) => any
  ? Arguments
  : never;
```